design/accepted/gc_plus/gc-plus-cloud.md

```diff
@@ -0,0 +1,52 @@
+# Managed Uncommitted Garbage Collection for lakeFS cloud - Execution Plan
+
+## Goals
+1. Run UGC for lakeFS cloud users
+2. Allow Cloud users to clean up storage space by deleting objects from the underlying storage for cost efficiency
+3. Support new and existing cloud users (both repository structures)
+
+## Non-Goals
+1. Remove lakeFS limitations during the UGC job run
+2. Improve UGC OSS support
+
+### User Story
+When creating a new cloud installation - on the GC initializations page,
+add UGC initialization (default run job=true for all repositories, default sweep = true).
+It can get the same rules as the GC.
+Add a link to the documentation costrains (lakeFS writes during the job, explain debugging and report).
+
+### Permissions
+Create default UGC user with permissions for every new cloud env (The permissions to Enigma was added to the GC user)
+- lakeFS permissions - prepare uncommitted gc
+- Bucket permissions - role to list, write, delete
+
+### Managed UGC job
+- Control plane interface (design with cloud native team)
+  - Add step for the create cloud installation
+  - UGC to get configuration for the spark job (configured role, secret and access key, installation endpoint, repository)
+  - User to configure min age seconds time
+  - Backup & Restore
+- Deployment of new UGC version
+- UGC client and server compatibility
+- Run the UGC periodically (AWS lambda cron job, Airflow DAG, EMR serverless, etc.)
+- Declare SLA (might require performance improvements)
+
+### Metrics and Logging
+- Add UGC logging and metrics to the cloud monitoring tools
+- Alerti configuration / on-call guide
+- Access to the cloud user metadata for debugging
+
+### Testing
+- Test UGC on scale
+  - (Enigma dev repo) Performance and correctness - mark only (WIP)
+    - Ensure only objects that are eligible for deletion are marked for deletion
+- Enigma production repo
+  - Old repository structure performance
+- Compatibility tests - UGC client and lakeFS server
+- Test sweep functionality
+- E2E tests for UGC cloud
+
+### Performance improvements
```

Comment:
I want to suggest to list the performance improvements explicitly by order of execution. I would start with:
1. Implement parallel listing for old repository structure (which will solve a major part of Enigma data)
2. Incremental run

Please note that each of these tasks requires execution of their own, performance measuring, testing etc.
Also, what should we do if our suggested improvements are not sufficient? What is our fail fast plan?

Reply:
Our fail fast is to test the UGC on scale (Enigma) before implementing it to the cloud. It may suggest problematic areas.

Reply:
This fail fast plan means that if UGC doesn't scale on Enigma's prod env, then we abandon the plan to add it to MGC. I don't think this is something we want to do.
On the other hand - we there is no clear and quick cutoff on where we need to stop and re-evaluate our efforts. The plan assumes we don't fail

Reply:
I'm not sure we need a fail fast plan on MUGC. It is not an unknown feature, if we decided we want a managed job for the cloud users why would we want to stop? I think we just need to start with a minimum managed job (no user control on the configuration, no backup) and then extend it will be needed